### ISSUE TITLE
[Question][Bug] How can i add Header on specific routes ? 

### ISSUE THREAD

**Issue by Horsty80** [microsoft/playwright#13527]
I want to add custom header like `"X-E2E-USER": "fake-user@e2e.com"` on header only for specific route (ie: i want to add header for request on my API)

I found this in the doc [https://playwright.dev/docs/api/class-route#route-continue](https://playwright.dev/docs/api/class-route#route-continue) that say i can overrides headers.

I try this but it's doesn't work. Headers are not overrides.

```js
await page.route(/\/api\//i, async (route, request) => {
    // Override headers
    const existingHeaders = await request.allHeaders();
    const headers = {
      ...existingHeaders,
      "X-E2E-USER": "fake-user@e2e.com", 
    };
    route.continue({ headers });
  });
```

Maybe i miss something, but it's not explain in the docs, maybe forget ? 

I don't want to use `setExtraHTTPHeaders` because it's override all http call, and i don't want change all my request

```js
  await page.setExtraHTTPHeaders({ "X-E2E-USER": "fake-user@e2e.com" });
```

**Comment by Horsty80** [comment]
Maybe issue is the same as explain on this [#13106](https://github.com/microsoft/playwright/issues/13106)

**Comment by aslushnikov** [comment]
@Horsty80 Just tried the following and it worked for me:

```js
// a.mjs
import { chromium } from '@playwright/test';

const browser = await chromium.launch();
const page = await browser.newPage();
await page.route('**/*', async (route, request) => {
  const headers = await request.allHeaders();
  route.continue({
    headers: {
      ...headers,
      'X-E2E-USER': 'fake-user@e2e.com',
    },
  });
});

await page.goto('https://postman-echo.com/get?foo1=bar1&foo2=bar2');
const text = await page.locator('body').textContent();
console.log(JSON.stringify(JSON.parse(text), null, 2));
await browser.close();
```

can you try running this snippet, does it work for you? If yes, would you mind creating a similar snippet that illustrates your issue?

**Comment by Horsty80** [comment]
@aslushnikov you snippet doesn't do thing differently but indeed if i request Postman echo page i see my header set with route.continue();

But how i can not see them when i try to log every request ? 

Basic snipet, only the getting started from playwright

```js
import { test, expect, Request } from "@playwright/test";

test("basic test", async ({ page }) => {
  page.on("request", async (request: Request) => {
    const headers = await request.allHeaders();
    console.log(">>", request.url(), ">>", headers);
  });
  await page.setExtraHTTPHeaders({ "EXTRA-HTTP-HEADER": "some-value" });
  await page.route("**/*", async (route, request) => {
    const headers = await request.allHeaders();
    route.continue({
      headers: {
        ...headers,
        "X-E2E-USER": "fake-user@e2e.com",
      },
    });
  });
  await page.goto("https://playwright.dev/");
  const title = page.locator(".navbar__inner .navbar__title");
  await expect(title).toHaveText("Playwright");
});

```

The `page.on("request")` is use to log / debug for this example.
The console output of the `console.log()` of the headers are this

```bash
>> https://playwright.dev/ >> {
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
  'accept-language': 'en-US',
  'extra-http-header': 'some-value',
  'upgrade-insecure-requests': '1',
  'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4929.0 Safari/537.36'
}
```

I see `'extra-http-header': 'some-value',` set by the command `await page.setExtraHTTPHeaders({ "EXTRA-HTTP-HEADER": "some-value" });` but i didn't see `"X-E2E-USER": "fake-user@e2e.com",` set bye `route.continue()` method.

That i exepect to see i don't understand why ?

**Comment by Horsty80** [comment]
What i try to achieve are :
- go to my page
- click on button -> that request my api
- add custom header on this specific request to my api 

What i don't want : 
- use `setExtraHTTPHeaders` because i don't need this custom header on all my route (that lead to some issue with some files)

**Comment by aslushnikov** [comment]
> But how i can not see them when i try to log every request ?

@Horsty80 the `page.on('request')` event happens **before** the route handler, so you get headers there **before** you modify them in `page.route`.

**Comment by Horsty80** [comment]
@aslushnikov yes `before` is the part is missing for understanding this behavior ^^

Indeed all the mock is working, it's just the `how` i check that isn't good.

Actually my test is good, i get indeed my custom header so i close this issue

Thanks for your answer :)

**Comment by ben-gooding-sky** [comment]
There is a workaround for this, to re-fetch the request object inside the `on('request'...`, like so:

`(await request.response())?.request().headers()`

Not very nice, but it does work and shows you correctly the extra added header 🤷
